src/htm/ntypes/Value.cpp

```diff
@@ -15,271 +18,703 @@
  * along with this program.  If not, see http://www.gnu.org/licenses.
  * --------------------------------------------------------------------- */
 
-/** @file
- * Implementation of the Value class
- */
-
+#include <htm/ntypes/BasicType.hpp>
 #include <htm/ntypes/Value.hpp>
 #include <htm/utils/Log.hpp>
 
+#include <algorithm> // transform
+#include <cerrno>
+#include <cstring> // std::strerror(errno)
+#include <iomanip>
+#include <iostream>
+#include <regex>
+#include <stack>
+
 using namespace htm;
 
-Value::Value(std::shared_ptr<Scalar> &s) {
-  category_ = scalarCategory;
-  scalar_ = s;
-}
+#define ZOMBIE_MAP ((size_t)-1) // Means the key of zombie was a map key
+#define ZOMBIE_SEQ 0            // Means the key of zombie was a seq key
```

Comment:
what is this zombie terminology?

Reply:
> what is this zombie terminology?

I tried to describe it where I define operator[ ].  I will expand on that description.
Consider some code that looks like this:
```
   Value vm;
   vm.parse("{ hello: something}");
   Value& node = vm["hello"];
   node = vm["newNode"];
   node = "new value to be inserted into the tree";
```
When we do the lookup with the vm["hello"] it exists so we get a reference to a node in the tree that matches the key.

However, if the lookup did not find anything (as is the case with vm["newNode"]) we don't have anything to return a reference to.  What STL does is create an entry in the tree and return that but I would rather not change the tree until there is a value being assigned to it.  So what I do is create a zombie node, assign the key value to it,  and return a reference to it. This is just a Value node of type Empty which is not linked into the tree but holds the key and a pointer to the parent in case it is assigned to.

When the operator= makes the assignment, it then puts the value into the zombie, and then adds the zombie to the parent in the tree using the key, changing its type to Scalar. At that point it is no longer a zombie.

The problem is where does that node exist until a value can be assigned?  The caller has only a reference to it.  So what I do is put the zombie object on the parent node in a shared_ptr.  The down side is that there can be only one outstanding lookup (or zombie) per parent node that can be assigned to. Its scope lasts until the next lookup on that node.

A complication is that the caller could be holding a reference to that zombie node. When I add the zombie node to the internal map, it adds a copy so the address changes.  So that is what the assigned_ pointers are for.   They redirect all operations on a zombie to the real node after it has been assigned to.  By doing some shared_ptr magic I think I can get rid of that extra re-direction.